### DISCUSSION TITLE
REST API for checking if a specific user is a sponsor and how much they are sponsoring

### DISCUSSION THREAD

**Discussion by lavender-sys** [community/community#54644] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Product Feedback

### Body

hiya! On my web app I want to give sponsors a little icon. Currently GitHub Sponsors only offers a webhook for this, which works, but it doesn't have an easy way of checking if a user is a sponsor for cases where maybe the server received the webhook but was unable to save the new data. I want to have a little refresh button in my web app's settings that when clicked uses some GitHub provided REST API to check if that specific user is sponsoring and for how much, but at the moment it doesn't seem like there's an API for this.

**Comment by octokatherine** [comment]
Hi @catgirlinspace! I'm an engineer on the Sponsors team. The data you're looking for is available through GitHub's GraphQL API: https://docs.github.com/en/graphql/reference/objects#sponsorship

**Reply by lavender-sys** [reply]
hi! I did see this while looking but wasn't sure how to use it exactly. I'm not experienced in using GraphQL APIs at all and wasn't sure how/if I could query by just a specific user id to check if they were sponsoring me. I think I'd have to query each page and then iterate over that to find the specific user I'm looking for?

**Reply by octokatherine** [reply]
The easiest way to find out if a user is sponsoring you would be to use the `isSponsoredBy` field on a User. For example:

```
query {
  viewer {
      isSponsoredBy(accountLogin: "octokatherine")
  }
}
```

this would tell you whether I am sponsoring you. You can test it out on the [GraphQL Explorer](https://docs.github.com/en/graphql/overview/explorer)

**Reply by lavender-sys** [reply]
Ah, didn't know I could do that, thanks!! Is there a way to also get back how much they're sponsoring for or the tier?

**Reply by octokatherine** [reply, marked as answer]
Yep! You could do something like this 

```
query {
  user(login:"octokatherine") {
    sponsorshipForViewerAsSponsorable(activeOnly:true) {
    	tier {
        name
        monthlyPriceInDollars
      }  
    }
  }
}
```

replacing my login with the sponsor's login

**Reply by lavender-sys** [reply]
Huh, didn't know GraphQL could do that! This looks like exactly what I need, thanks so much!!!!

**Reply by QROkes** [reply]
Hi @octokatherine I know this is an "old" thread.  Can I ask you one question?

Considering `sponsorshipForViewerAsSponsorable(activeOnly:true)`, one-time donors are considered as active? For how long?

Thanks in advance!

**Reply by octokatherine** [reply]
Hi @QROkes! One time sponsors are active for 30 days.

**Reply by honzajavorek** [reply]
The query, copy-pasted as is and login replaced, gives me only
```json
{
  "data": {
    "user": {
      "sponsorshipForViewerAsSponsorable": null
    }
  }
}
```
Not sure what's wrong or if the API has changed 🤔

**Reply by antn** [reply]
@honzajavorek Hi, I'm from the Sponsors engineering team!

Is the user you're accessing that field for an active sponsor for you? That field defaults to only displaying data for active sponsorships. If you want to view their previous sponsorship, you can change the query to:

```gql
query {
  user(login: "antn") {
    sponsorshipForViewerAsSponsorable(activeOnly: false) {
      tier {
        name
        monthlyPriceInDollars
      } 
    }
  }
}
```

**Reply by honzajavorek** [reply]
I have both active and inactive sponsors, but neither were listed. I've seen this behavior both in the explorer and as a real response from the API using a fine-grained personal token. I thought I'd solve what I needed with something like this:
```graphql
query fetch($login: String!) {
  user(login: $login) {
    sponsorshipsAsMaintainer(
      first: 100
      includePrivate: false
      activeOnly: false
    ) {
      nodes {
        isActive
        sponsorEntity {
          ... on User {
            name
            login
            avatarUrl
            url
          }
          ... on Organization {
            name
            login
            avatarUrl
            url
          }
        }
      }
    }
  }
}

```
As I want to list the sponsors on a website, it's important for me to filter out the ones who are private. Either manually later or in the API, but I need the information. That's why some of the ways where I can get sponsors from the API don't fit my use case, the info whether it's private isn't there. The query above fits the requirement.

Then, I'd perform another query to get the amounts and pair the data in my Python app:
```graphql
query fetch($login: String!) {
  user(login: $login) {
    lifetimeReceivedSponsorshipValues(first: 100, orderBy: {field: LIFETIME_VALUE, direction: DESC}) {
      nodes {
        amountInCents
        sponsor {
          ... on User { login }
          ... on Organization { login }
        }
      }
    }
  }
}
```
However, this didn't work either. Well it worked and I've been getting data exactly up until the moment I asked the API about anything regarding the amounts, even just the tier name. In those cases the API has turned me down saying I have no permissions to ask for those with a fine-grained personal token. It didn't say what I need and the GitHub tokens and permissions feel like a maze to me, so I gave it a few tries and then gave up. Given the wording I assume it's just not possible with a PAT. In overall, the experience to „just get a list of my sponsors“ feels like a minefield - something I thought will take 5mins to figure out took me a whole day.

I'm asking the API about my own data, not someone else's, so I don't understand this limitation, but whatever. I just wanted to list my sponsors on my site, putting the ones paying most as first in the list. My life's too short for playing with this API, so I'll just list them and sort them randomly 😄

**Reply by antn** [reply]
Very sorry about the frustration here, @honzajavorek!

`sponsorshipsAsMaintainer` is a good field for getting your sponsorships, like you mentioned. You can also access the `privacyLevel` field on each node to see if it is a public or private sponsorship, if that helps you filter them out on your end.

As for the issue with fetching `lifetimeReceivedSponsorshipValues` with a fine-grained token – it does look like there is a bug here that prevents that field from being accessed by a fine-grained token. It also has a bug where it doesn't properly tell you what scopes/permissions are missing.

We've opened an issue internally for this, and will investigate how we can fix that. I am very sorry for the trouble!

As a workaround for now, you can use a classic personal access token instead of a fine-grained one, and grant it the `user` scope. That should allow you to fetch this information without any issues. Hopefully that can help, if you want to give it a try!

**Reply by honzajavorek** [reply]
Wow, that's awesome! I didn't expect this outcome 😄 It didn't occur to me it could be a bug, I just thought you're super secure about the actual numbers. Thank you.
